File: middle/leetcode_0039.py
"""
    1、目标数据给定，组合数据个数不确定
    2、回溯法（还没有搞清楚这个方法的核心思想）
"""

class Solution:
    def combinationsSum(self, candidates, target):
        """
        回溯法，层层递减，得到符合条件的路径就加入到结果中，超出则剪枝
        主要是要注意一些细节，避免重复等
        """
        size = len(candidates)
        if size <= 0:
            return []

        # 先排序，便于后面剪枝
        candidates.sort()

        path = []
        res = []
        self._find_path(self, target, path, res, candidates, 0, size)
        return res

    def _find_path(self, target, path, res, candidates, begin, size):
        """沿着路径往下走"""
        if target == 0:
            res.append(path.copy())
        else:
            for i in range(begin, size):
                left_num = target - candidates
                # 如果剩余值为负数，说明超过了，剪枝
                if left_num < 0:
                    break
                # 否则把当前值加入路径
                path.append(candidates[i])
                # 为避免重复解，我们把比当前值小的参数也从下一次寻找中剔除
                # 因为根据他们得到的解一定在之前就找到过了
                self._find_path(left_num, path, res, candidates, i, size)
                # 记得把当前值移出路径才能进入下一个值的路径
                path.pop()

File: CPP_Solution/middle/leetcode_0039.cpp
// 2022/11/23  author:WH 组合总和
#include<iostream>
#include<vector>
#include<algorithm> //sort函数的使用
using namespace std;

class Solution{
private:
    vector<vector<int>> result;
    vector<int> path;
    void backtracking(vector<int>& candidates, int target, int sum, int startIndex){
        if (sum == target){
            result.push_back(path);
            return;
        }
        // 剪枝：如果sum+candidates[i] > target 就终止遍历
        for (int i = startIndex; i < candidates.size() && sum + candidates[i] <= target; i++){
            sum += candidates[i];
            path.push_back(candidates[i]);
            backtracking(candidates, target, sum, i);
            sum -= candidates[i];
            path.pop_back();
        }
}

public:
    vector<vector<int>> combinationsSum(vector<int>& candidates, int target){
        result.clear();
        path.clear();
        sort(candidates.begin(), candidates.end()); //对candidates排序
        backtracking(candidates, target, 0, 0);
        return result;
    }